### DISCUSSION TITLE
How does the CLR confirm the validity of pointer-like types?

### DISCUSSION THREAD

**Discussion by atynagano** [dotnet/runtime#53665] [Q&A]
C# has some pointer-like type such as pointers, refs, and Span. But how these types avoid pointers getting invalidated due to memory relocation by GC. I think it is common to use the fixed statement for pinning to acquire array references. However, ref, Span, Unsafe.AsPointer(), etc. do not perform pinning. Is the pinning actually done, or does the relocation sometimes cause an error?

I found an answer that seems relevant, but even if that is the case I don't know when or how it is performed.
https://stackoverflow.com/questions/47419240/how-does-the-c-sharp-garbage-collector-find-objects-whose-only-reference-is-an-i

**Comment by huoyaoyuan** [comment]
> But how these types avoid pointers getting invalidated due to memory relocation by GC.

Pointers are unknown to GC and can be leaked into unmanaged code. Thus obtaining a pointer requires pinning.

`ref`s functions like pointer, but it's known by GC, like object reference. It can only live in managed code, so that GC has the information and right to reposition it.

> Is the pinning actually done, or does the relocation sometimes cause an error?

Neither. It doesn't require pinning, like normal object reference.
